fendermint/vm/interpreter/src/chain.rs

```diff
@@ -214,39 +223,67 @@ where
                     }
 
                     // We can use the same result type for now, it's isomorphic.
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDownExec(_) => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownExec(p) => {
+                    let validator_set =
+                        atomically_or_err::<_, fendermint_vm_topdown::Error, _>(|| {
+                            provider.validator_set(p.height as u64)
+                        })
+                        .await?
+                        .ok_or_else(|| {
+                            anyhow!("cannot find validator set for block: {}", p.height)
+                        })?;
+
+                    let finality = IPCParentFinality {
+                        height: p.height as u64,
+                        block_hash: p.block_hash,
+                    };
+                    let msg = parent_finality_to_fvm(finality.clone(), validator_set)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+
+                    atomically_or_err::<_, fendermint_vm_topdown::Error, _>(|| {
+                        provider.set_new_finality(finality.clone())
+                    })
+                    .await?;
+
+                    // TODO: execute top down messages,
+                    // TODO: see https://github.com/consensus-shipyard/fendermint/issues/241
```

Comment:
Thank you for the link. Minor but it belongs further up _before_ `set_new_finality`, because by this point the cross messages would have disappeared form the cache. And you also need the previous finality for it, which should also be in the provider before it's overwritten. 

And here I see the value in the range query available for cross messages, although I still think that it would help if the block_hash and validator_set could not individually return `None`.

Reply:
An example of why the current design is not ideal:
* `provider.validator_set(p.height as u64)` would return `None` if it's not found, which you handle by crashing Fendermint by returning ` anyhow!("cannot find validator set for block: {}", p.height)`. 
* `provider.cross_messages(from, to)` return `Vec<CrossMsg>` so it would be empty, if you implemented this execution where the TODO is now, you would never see an error, it would look like it's all dandy, just there were no messages. 

While in reality both should fail because of the same reason.

Reply:
They all are simplified, they will not return `Option` but instead exponentially backoff retry or util it fails.